File: mobile/lib/utils/skill_icons.dart
import 'package:http/http.dart' as http;

/// Local fallback when API [iconUrl] is unavailable.
/// Prefer `/api/reference/tech-stack` search or resolve from [ReferenceApi].
const _skillIconsBase = 'https://skillicons.dev/icons?i=';

final Map<String, String> _skillIconSvgCache = {};

const _slugOverrides = <String, String>{
  'typescript': 'ts',
  'javascript': 'js',
  'node.js': 'nodejs',
  'nodejs': 'nodejs',
  'c++': 'cpp',
  'c#': 'cs',
  'vue.js': 'vue',
  'vue': 'vue',
  'next': 'nextjs',
  'next.js': 'nextjs',
  'tailwind': 'tailwind',
  'tailwindcss': 'tailwind',
  'postgres': 'postgresql',
  'postgresql': 'postgresql',
  'go': 'golang',
  'golang': 'golang',
  'k8s': 'kubernetes',
  'aws': 'aws',
  'gcp': 'gcp',
  'figma': 'figma',
  'git': 'git',
  'github': 'github',
  'gitlab': 'gitlab',
  'docker': 'docker',
  'mongodb': 'mongodb',
  'redis': 'redis',
  'html': 'html',
  'css': 'css',
  'react': 'react',
  'angular': 'angular',
  'python': 'python',
  'java': 'java',
  'rust': 'rust',
  'php': 'php',
  'ruby': 'ruby',
  'swift': 'swift',
  'kotlin': 'kotlin',
  'terraform': 'terraform',
  'linux': 'linux',
  'graphql': 'graphql',
  'prisma': 'prisma',
  'vite': 'vite',
  'express': 'express',
  'nestjs': 'nestjs',
  'jest': 'jest',
  'cypress': 'cypress',
};

String skillIconSlug(String displayName) {
  final trimmed = displayName.trim();
  if (trimmed.isEmpty) return '';
  final lower = trimmed.toLowerCase();
  final noSpaces = lower.replaceAll(RegExp(r'\s+'), '');
  final noDots = noSpaces.replaceAll('.', '');
  return _slugOverrides[noDots] ??
      _slugOverrides[lower] ??
      _slugOverrides[noSpaces] ??
      noDots;
}

String skillIconUrl(String displayName) {
  final slug = skillIconSlug(displayName);
  return slug.isEmpty ? '' : '$_skillIconsBase${Uri.encodeComponent(slug)}';
}

String skillIconUrlBySlug(String slug) {
  final s = slug.trim();
  return s.isEmpty ? '' : '$_skillIconsBase${Uri.encodeComponent(s)}';
}

bool isSkillIconSvgUrl(String url) {
  final lower = url.toLowerCase();
  return lower.contains('skillicons.dev') || lower.endsWith('.svg');
}

/// skillicons.dev wraps each icon in an outer SVG with a nested inner SVG.
/// [flutter_svg] cannot render that wrapper, so we extract the inner document.
String flattenSkilliconsSvg(String raw) {
  final trimmed = raw.trim();
  final lower = trimmed.toLowerCase();
  final start = lower.lastIndexOf('<svg');
  if (start < 0) return trimmed;

  final end = lower.indexOf('</svg>', start);
  if (end < 0) return trimmed;

  return trimmed.substring(start, end + 6);
}

/// Fetch and flatten a remote skillicons SVG for [SvgPicture.string].
Future<String?> fetchSkillIconSvg(String url) async {
  final key = url.trim();
  if (key.isEmpty) return null;

  final cached = _skillIconSvgCache[key];
  if (cached != null) return cached;

  try {
    final res = await http.get(Uri.parse(key)).timeout(const Duration(seconds: 8));
    if (res.statusCode < 200 || res.statusCode >= 300) return null;
    final flat = flattenSkilliconsSvg(res.body);
    _skillIconSvgCache[key] = flat;
    return flat;
  } catch (_) {
    return null;
  }
}

File: webapp/src/lib/profile/skillIcons.ts
const SKILL_ICONS_BASE = "https://skillicons.dev/icons?i=";
const SLUG_OVERRIDES: Record<string, string> = {
  typescript: "ts",
  javascript: "js",
  "node.js": "nodejs",
  nodejs: "nodejs",
  "c++": "cpp",
  "c#": "cs",
  "vue.js": "vue",
  vue: "vue",
  next: "nextjs",
  "next.js": "nextjs",
  tailwind: "tailwind",
  tailwindcss: "tailwind",
  postgres: "postgresql",
  postgresql: "postgresql",
  go: "golang",
  golang: "golang",
  k8s: "kubernetes",
  aws: "aws",
  gcp: "gcp",
  figma: "figma",
  git: "git",
  github: "github",
  gitlab: "gitlab",
  docker: "docker",
  mongodb: "mongodb",
  redis: "redis",
  html: "html",
  css: "css",
  react: "react",
  angular: "angular",
  python: "python",
  java: "java",
  rust: "rust",
  php: "php",
  ruby: "ruby",
  swift: "swift",
  kotlin: "kotlin",
  terraform: "terraform",
  linux: "linux",
  graphql: "graphql",
  prisma: "prisma",
  vite: "vite",
  express: "express",
  nestjs: "nestjs",
  jest: "jest",
  cypress: "cypress",
};
export function getSkillIconSlug(displayName: string): string {
  const trimmed = displayName.trim();
  if (!trimmed) return "";
  const lower = trimmed.toLowerCase();
  const noSpaces = lower.replaceAll(/\s+/g, "");
  const noDots = noSpaces.replaceAll(".", "");
  return (
    SLUG_OVERRIDES[noDots] ??
    SLUG_OVERRIDES[lower] ??
    SLUG_OVERRIDES[noSpaces] ??
    noDots
  );
}
export function getSkillIconUrl(displayName: string): string {
  const slug = getSkillIconSlug(displayName);
  return slug ? `${SKILL_ICONS_BASE}${encodeURIComponent(slug)}` : "";
}
export function getSkillIconUrlBySlug(slug: string): string {
  const s = slug?.trim();
  return s ? `${SKILL_ICONS_BASE}${encodeURIComponent(s)}` : "";
}
export function getSkillIconsCombinedUrl(displayNames: string[]): string {
  const slugs = displayNames.map(getSkillIconSlug).filter(Boolean);
  return slugs.length
    ? `${SKILL_ICONS_BASE}${slugs.map(encodeURIComponent).join(",")}`
    : "";
}
export function preloadTechStackItems(
  items: ReadonlyArray<{
    name: string;
    iconUrl?: string;
    iconSlug?: string;
  }>,
): void {
  if (typeof window === "undefined" || items.length === 0) return;
  const slugs = items
    .map((item) => {
      const fromSlug = item.iconSlug?.trim();
      if (fromSlug) return fromSlug;
      return getSkillIconSlug(item.name);
    })
    .filter(Boolean);
  const combined = slugs.length
    ? `${SKILL_ICONS_BASE}${slugs.map(encodeURIComponent).join(",")}`
    : "";
  if (combined) {
    const batch = new window.Image();
    batch.decoding = "async";
    batch.src = combined;
  }
  for (const item of items) {
    const url = resolveTechStackIconSrc(item);
    if (!url) continue;
    const img = new window.Image();
    img.decoding = "async";
    img.src = url;
  }
}
export function resolveTechStackIconSrc(
  item: Readonly<{
    name: string;
    iconUrl?: string;
    iconSlug?: string;
  }>,
): string {
  const fromApi = item.iconUrl?.trim();
  if (fromApi) return fromApi;
  const slug = item.iconSlug?.trim() || getSkillIconSlug(item.name);
  return slug ? getSkillIconUrlBySlug(slug) : getSkillIconUrl(item.name);
}
export function preloadSkillIcons(displayNames: string[]): void {
  if (typeof window === "undefined" || displayNames.length === 0) return;
  const combined = getSkillIconsCombinedUrl(displayNames);
  if (combined) {
    const batch = new window.Image();
    batch.decoding = "async";
    batch.src = combined;
  }
  for (const name of displayNames) {
    const url = getSkillIconUrl(name);
    if (!url) continue;
    const img = new window.Image();
    img.decoding = "async";
    img.src = url;
  }
}
